File: Restaurant/app/src/main/java/com/example/restaurant/data/model/Reservation.kt
package com.example.restaurant.data.model

import com.google.gson.annotations.SerializedName
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Data class representing a reservation in the restaurant
 */
data class Reservation(
    @SerializedName("ReservationID")
    val id: Int,
    @SerializedName("ReservationCode")
    val code: String?,
    @SerializedName("UserID")
    val userId: Int,
    @SerializedName("FullName")
    val fullName: String,
    @SerializedName("Phone")
    val phone: String,
    @SerializedName("TableID")
    val tableId: Int?,
    @SerializedName("GuestCount")
    val guestCount: Int,
    @SerializedName("ReservationDate")
    val reservationDate: String,
    @SerializedName("Status")
    val status: String,
    @SerializedName("Note")
    val note: String?,
    @SerializedName("CreatedAt")
    val createdAt: String?,
    @SerializedName("UpdatedAt")
    val updatedAt: String?,
    @SerializedName("CheckInTime")
    val checkInTime: String?,
    @SerializedName("CheckOutTime")
    val checkOutTime: String?,
    @SerializedName("reservationItems")
    val reservationItems: List<ReservationItem>? = null,
    @SerializedName("payment")
    val payment: Payment? = null,
    @SerializedName("table")
    val table: Table? = null
) {
    fun getFormattedDate(): String {
        try {
            val inputFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
            val outputFormat = SimpleDateFormat("dd/MM/yyyy HH:mm", Locale.getDefault())
            val date = inputFormat.parse(reservationDate) ?: return reservationDate
            return outputFormat.format(date)
        } catch (e: Exception) {
            return reservationDate
        }
    }
    
    fun getStatusColor(): Int {
        return when(status) {
            "Đã xác nhận" -> android.graphics.Color.GREEN
            "Chờ xác nhận" -> android.graphics.Color.YELLOW
            "Đã hủy" -> android.graphics.Color.RED
            "Đã hoàn tất" -> android.graphics.Color.BLUE
            else -> android.graphics.Color.GRAY
        }
    }
    
    fun getTotalAmount(): Double {
        return reservationItems?.sumOf { it.price * it.quantity } ?: 0.0
    }
} 

File: hari-restaurant/app/Models/Reservation.php
<?php


namespace App\Models;


use App\Models\user\Cart;
use App\Models\user\MenuItem;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Factories\HasFactory;


class Reservation extends Model
{
    use HasFactory;


    protected $table = 'reservation';
    protected $primaryKey = 'ReservationID';


    const CREATED_AT = 'CreatedAt';
    const UPDATED_AT = 'UpdatedAt';


    protected $fillable = [
        'ReservationCode',
        'UserID',
        'FullName',
        'Phone',
        'TableID',
        'GuestCount',
        'ReservationDate',
        'Status',
        'Note',
        'CheckInTime',
        'CheckOutTime'
    ];


    protected $dates = [
        'ReservationDate',
        'CheckInTime',
        'CheckOutTime',
        'CreatedAt',
        'UpdatedAt'
    ];


    public function user()
    {
        return $this->belongsTo(User::class, 'UserID', 'id');
    }


    public function table()
    {
        return $this->belongsTo(Table::class, 'TableID', 'TableID');
    }


    public function reservationItems()
    {
        return $this->hasMany(ReservationItem::class, 'ReservationID', 'ReservationID');
    }


    public function cartItems()
    {
        return $this->hasMany(Cart::class, 'ReservationID', 'ReservationID');
    }


    public function menuItems()
    {
        return $this->hasManyThrough(
            MenuItem::class,
            Cart::class,
            'ReservationID', // Khóa ngoại trên bảng cart
            'ItemID', // Khóa chính trên bảng menuitem
            'ReservationID', // Khóa chính trên bảng reservation
            'item_id' // Khóa ngoại trên bảng cart tham chiếu đến menuitem
        );
    }


    public function payment()
    {
        return $this->hasOne(Payment::class, 'ReservationID', 'ReservationID');
    }


    public function setCheckInTimeAttribute($value)
    {
        $this->attributes['CheckInTime'] = $value ? \Carbon\Carbon::parse($value) : null;
    }


    public function setCheckOutTimeAttribute($value)
    {
        $this->attributes['CheckOutTime'] = $value ? \Carbon\Carbon::parse($value) : null;
    }
}
